### ISSUE TITLE
WHat is the proper way to reconnect?

### ISSUE THREAD

**Issue by alexiordan** [aspnet/SignalR#1611]
Since reconnect is gone, when using WebSocket as transport layer what is the advisable way for attempting to maintain a connection always on (by attempting reconnection when connection drop is detected)?

It is more like a documentation issue since I didn't find it in documentation.

On **server side** I have set a KeepAlive of 3 seconds. I have also seen a NegotiateTimeout property. What is it for?
```
services.AddSignalR(
                options =>
                {
                    options.KeepAliveInterval = TimeSpan.FromSeconds(3);
                    //What is this for?
                    //options.NegotiateTimeout
                });
```
On **client side** I have requested for WebSockets with KeepAlive 3 seconds.
```
 HubConnection = new HubConnectionBuilder()
                    .WithTransport(Microsoft.AspNetCore.Sockets.TransportType.WebSockets)
                    .WithWebSocketOptions(a => a.KeepAliveInterval = TimeSpan.FromSeconds(3))
                    .Build();
```
There is also ServerTimeout option. What is it for?
```
HubConnection.ServerTimeout = TimeSpan.FromSeconds(9);
```
I am currently handling the closing event and performing following actions:
```  
await HubConnection.StopAsync();
await Task.Delay(2000);
 HubConnection = new HubConnectionBuilder()
                    .WithTransport(Microsoft.AspNetCore.Sockets.TransportType.WebSockets)
                    .WithWebSocketOptions(a => a.KeepAliveInterval = TimeSpan.FromSeconds(3))
                    .Build();
//subscribe to events
await HubConnection.StartAsync();
```

Is this the correct way to do it, by building a new HubConnection on closing?

Before StopAsync() and starting it again I placed a 2 seconds delay. Should I keep it?

**Comment by anurse** [comment]
> I have also seen a NegotiateTimeout property.

After establishing the connection, the client is expected to send a message indicating the protocol it wishes to use. The `NegotiateTimeout` is how long the server will wait for that message to be sent. The default value should be totally fine for most, if not all, use cases.

> There is also ServerTimeout option. What is it for?

If the client hasn't heard from the server within the `ServerTimeout`, it will close the connection and report an error indicating that the server is not present. This should always be significantly larger than the `KeepAliveInterval` on the server. The `KeepAliveInterval` on the server configures the server to send a keep-alive message within that interval if no other messages have been sent.

> I am currently handling the closing event and performing following actions:

You actually don't have to re-create the `HubConnection`. You can simply call `StartAsync` on it again. The reason this flow is manual is that it is very likely you will want to write application logic for the interval between disconnect and reconnect, especially since SignalR no longer buffers messages for you. You could display a message to a user or disable some UI, for example.

> Before StopAsync() and starting it again I placed a 2 seconds delay. Should I keep it?

This depends a lot on what you see in the wild. For example, if the server has 1000 clients connected and it goes down, when it comes back, it'll have 1000 clients immediately trying to reconnect (this was known as a "Reconnect Storm" in SignalR for ASP.NET). You should be prepared for this scenario and see if you can put in some diagnostics to understand client behavior when this happens. You could have clients pick a random delay between when the disconnect occurs and when reconnection happens to help mitigate this.

**Comment by alexiordan** [comment]
OK, now if I don't re-create the HubConnection I get the following behavior:
On Closed Event I get this:
```
Exception thrown: 'System.TimeoutException' in System.Private.CoreLib.dll
Server timeout (9000.00ms) elapsed without receiving a message from the server.
```
In the same Closed handler I attempt to reconnect like this:
```
 m_Logger.WriteLine<SignalRClient>("1");
await HubConnection.StopAsync();
m_Logger.WriteLine<SignalRClient>("2");
await Task.Delay(2000);
m_Logger.WriteLine<SignalRClient>("3");
await StartConnectionAsync();
m_Logger.WriteLine<SignalRClient>("4");
```

I only get logged: 
```
1
4
```

**Comment by alexiordan** [comment]
Thank you for your time.
Regarding the StopAsync. OK, I'll remove it and see how it works.

Regarding the parameters:
I am just trying to use the same parameters i did with the previous version of SignalR and I was pleased with the results.

One more question. What are the differences between setting the KeepAlive parameter on client-side vs setting it on the server side?

**Comment by anurse** [comment]
The only Keep-Alive that SignalR provides is from Server-to-Client. That is, a SignalR server will ensure that at least one message is sent within the Keep-Alive interval, sending a Ping message if there are no user messages to send.

The "client" Keep-Alive you've configured is part of the options provided to the WebSocket and refers to sending WebSocket [Ping](https://tools.ietf.org/html/rfc6455#section-5.5.2) frames. Those Ping frames are not visible to the server (they are handled at a lower level) so they have no impact on SignalR's behavior. Those frames only serve to ensure that HTTP proxies sitting in the middle of the connection don't terminate the connection, thinking the other side has timed out. Those frames are sent in both directions (by the server and by the client) automatically, and SignalR has no control over that except to expose the option to configure the duration. Honestly, I think there's very little value (and a cost of increased network traffic) to changing that value from it's default.

**Comment by alexiordan** [comment]
OK,
Not creating the HubConnection each time, I get eventually into this:
```
Cannot start a connection that is not in the Disconnected state.
```

1. I think when I get into this I should re-build the HubConnection. Is this correct?
2. If so, how can I know when should I rebuild the HubConnection and when not to? Can I query its state in some way?

**Comment by anurse** [comment]
Can you confirm what version of SignalR you are using? The preview 1 version should allow you to re-connect without requiring you to rebuild the `HubConnection`.

**Comment by alexiordan** [comment]
1. My version is: 1.0.0-preview1-final
2. So that message means that I should rebuild the HubConnection?
3. I'll try to reproduce and write you the steps.

**Comment by anurse** [comment]
> So that message means that I should rebuild the HubConnection?

No, it should be working without rebuilding the HubConnection. Versions **older** than preview 1 did require that you rebuild the connection.

> I'll try to reproduce and write you the steps.

That would help a lot. If you're able to narrow down a simple repro we can more easily identify the issue.

**Comment by BrennanConroy** [comment]
Looks like you're stackoverflowing
`StartConnectionAsync -> OnConnectionExceptionAsync -> OnConnectionClosedAsync -> StartConnectionAsync -> OnConnectionExceptionAsync -> etc`

**Comment by alexiordan** [comment]
It is indeed.
I have now changed it to 
```
private async Task StartConnectionAsync()
        {
            try
            {
                if (m_ConnectionIsStarted == false)
{
                    CreateHubConnection();
                }
             await   HubConection.StartAsync();
                m_ConnectionIsStarted = true;
            }
           
            catch (Exception ex)
            {
                await OnConnectionExceptionAsync(ex);
            }
        }

        private async Task OnConnectionExceptionAsync(Exception exception)
        {
            m_ConnectionIsStarted = false;
            await CleanUpHubConnectionAsync();
            await OnConnectionClosedAsync(exception);
        }

 private async Task CleanUpHubConnectionAsync()
        {
            if (HubConnection != null)
            {
                try
                {
                    await HubConnection.StopAsync();
                }
                catch
                {
                }


                if (HubConnection != null
{
                    try
                    {
                        HubConnection.Closed -= HubConnection_Closed;
                        await HubConnection.DisposeAsync();
                    }
                    catch
                    {
                    }
                    HubConnection = null;
                }
            }
        }

```
and let it run. 
Do you think should be better now?
I Debug Print on almost each line (removed in the code here) so hopefully will have something later.

**Comment by alexiordan** [comment]
Sorry! I now realize that updated approach is full of cycles again (theresuch I got into an OutOfMemoryException this time :-) )

So, back to my initial question: What is the proper way to reconnect?

I can try to set a timer and on Elapsed event to evaluate a variable set on ClosedHandler or on exception catch of the StartAsync. Upon evaluation to StartAsync or do nothing until next Elapsed event.

I think is a common scenario for the client to attempt reconnecting to server...

**Comment by anurse** [comment]
You should be able to simply call `HubConnection.StartAsync()` from inside the `HubConnection.Closed` event handler.

What is the primary problem you're seeing? From the looks of it, if you're seeing `OutOfMemoryException` or `StackOverflowException` errors in your code, it would appear that you are retrying indefinitely and continuing to encounter errors. What's the exception that's occurring during `StartAsync` to cause the retry loop? It's possible it's an unrecoverable error some kind.

In general, infinite retry is like an infinite loop, if a persistent error develops (network connectivity issues, etc.) you'll retry forever. You need to limit retries and at a certain point you have to give up and report the error to the user.

**Comment by anurse** [comment]
> Stopped the server for the entire test sesssion - never restarted

I see that you're testing the behavior when the server is unavailable. We shouldn't be leaking memory using the pattern I described (calling `StartAsync` in the `Closed` event handler). If you see `OutOfMemoryException` errors there, we should loop at getting more information. If you're able to capture a process dump when the `OutOfMemoryException` is thrown, I can help identify what objects are filling up memory to help identify the source of the leak.

**Comment by alexiordan** [comment]
Unfortunately I've just stopped the debugging session.
It took me 4 hours to reproduce it so I think tommorow, maybe, I'll be able to provide you the process 
dump. 
But, at first, let's see if we are on the same wavelength.

It is not the Closed event handler that I am trying to connect when the exception fires.

The Closed event is fired when the server goes offline after the client has been connected (once).

For every one of the next connecting attemtps, I catch the exception thrown by StartAsync() method.

Now, in this catch i tried 2 ways of reconnecting with behaviors reported above:

1. Just reconnecting by calling StartAsync() - as described in post [withoutCleanup](https://github.com/aspnet/SignalR/issues/1611#issuecomment-374449045) eventually getting into 
```
An unhandled exception of type 'System.Runtime.InteropServices.SEHException' occurred in System.Private.CoreLib.dll
External component has thrown an exception.
```
2. Attempt to clean the HubConnection object, rebuild it and afterwards calling StartAsync() - as described in post [withCleanup](https://github.com/aspnet/SignalR/issues/1611#issuecomment-374492221) eventually getting into
```
Exception thrown: 'System.OutOfMemoryException' in System.Private.CoreLib.dll
```

**Comment by anurse** [comment]
> Just reconnecting by calling StartAsync() - as described in post withoutCleanup eventually getting into

If you are retrying from a failed `StartAsync` call, this is the right approach, you should just try calling `StartAsync` again. That exception message is very odd, do you have any other information about it? (Stack Trace, etc.) Could you capture a process dump from when the exception occurs?

> Attempt to clean the HubConnection object, rebuild it and afterwards calling StartAsync()

This shouldn't be necessary, as I said earlier, but it also shouldn't be blowing out your memory. This is another place where a process dump would be useful.

**Comment by alexiordan** [comment]
OK, 
I'm happy I was not doing it completely wrong then. @BrennanConroy's [post ](https://github.com/aspnet/SignalR/issues/1611#issuecomment-374454705) make sense anyway.

I'll get back with dumps, but for now I have saved some StackTraces of the exceptions I caught just before the crashes.

I have logged the StackTraces for both cases. They are a little bit long. I don't know how to post them elseways. 
1. StackTrace for 
>  Just reconnecting by calling StartAsync() 
```
Server timeout (9000.00ms) elapsed without receiving a message from the server.

   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Sockets.Client.HttpConnection.<StopAsyncCore>d__55.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Sockets.Client.HttpConnection.<StartAsync>d__40.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.<StartAsyncCore>d__26.MoveNext()
... 1 340 times
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.<StartAsync>d__23.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at myApp.SignalRClient.<StartConnectionAsync>d__35.MoveNext()

The thread 0x634 has exited with code 0 (0x0).
The thread 0x1580 has exited with code 0 (0x0).

Exception thrown: 'System.Runtime.InteropServices.COMException' in System.Private.CoreLib.dll
WinRT information: A connection with the server could not be established

Exception thrown: 'System.Runtime.InteropServices.COMException' in System.Net.WebSockets.Client.dll
WinRT information: A connection with the server could not be established

Exception thrown: 'System.Runtime.InteropServices.COMException' in System.Private.CoreLib.dll
WinRT information: A connection with the server could not be established

An unhandled exception of type 'System.Runtime.InteropServices.SEHException' occurred in System.Private.CoreLib.dll
External component has thrown an exception.

```
2. StackTrace for 
> Attempt to clean the HubConnection object, rebuild it and afterwards calling StartAsync()
Although for the OutOfMemory the StackTrace I am not sure how relevant it is.
```
Unable to connect to the remote server

   at System.Net.WebSockets.WebSocketHandle.<ConnectAsyncCore>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.ClientWebSocket.<ConnectAsyncCore>d__16.MoveNext()
... 6 times
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.<StartAsync>d__23.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at myApp.SignalRClient.<StartConnectionAsync>d__35.MoveNext()
Exception thrown: 'System.OutOfMemoryException' in System.Private.CoreLib.dll
```

**Comment by anurse** [comment]
> An unhandled exception of type 'System.Runtime.InteropServices.SEHException' occurred in System.Private.CoreLib.dll
> External component has thrown an exception.

This one is the most suspicious one. The others are all the connection failure exceptions I'd expect to see if you have the server off for this test. There's no stack trace or anything providing any more context on it.

**Comment by anurse** [comment]
If you can inspect the [`ErrorCode`](https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.externalexception.errorcode(v=vs.110).aspx) property of the `SEHException` that's being thrown, that would help. I believe you can probably get it from the [`UnhandledException`](http://msdn.microsoft.com/en-us/library/windows/apps/xaml/windows.ui.xaml.application.unhandledexception.aspx) event on `Windows.UI.Xaml.Application`

**Comment by anurse** [comment]
Have you been able to use the `UnhandledException` event to gather any more information on the issue? Figuring out where the exception is coming from will help us narrow down exactly what's wrong. It looks like you've got the necessary `catch` blocks in place, so I don't know why it would be escaping to the `UnhandledException` event.

Are you also catching exceptions around your calls to `SendAsync`/`InvokeAsync`/etc.? Those could throw these exceptions as well, if the connection is down and you try to send a message.

**Comment by alexiordan** [comment]
1. Status:
After I talked to you I tried to reproduce it for 2 days in a row.
1st day - did not reproduce
2nd day - got a System Out of Memory, but the scenario is not valid. HubConnection was rebuild each time it tried to connect WITHOUT prior HubConnection disposal/cleanup.
For the exception on the 2nd day I have the memory dump. It is about 200 MB in size.

2. UnhandledException - I will used that too. Currently I only set Handled on true. I will have the exception printed on that one too.

3. The SEHException. I will keep trying to reproduce it.

**Comment by davidfowl** [comment]
The OOM might be https://github.com/aspnet/SignalR/pull/1104.

**Comment by alexiordan** [comment]
Hi everyone.
I've got again the 
> Cannot start a connection that is not in the Disconnected state.

No UnhandledException caught.

**Scenario:**
I had online both the server and client for about 3 hours.
Then I took the server down for about 7 hours while client was running.
After the 7 hours, I started the server and the mentioned exception appeared.

I was using the code that was:
>1. Just reconnecting by calling StartAsync()

and not the one that:

>2. Attempt to clean the HubConnection object, rebuild it and afterwards calling StartAsync()

I did not stop any of the processes so if you want me to provide any additional info, hopefully I can do it.

**Comment by alexiordan** [comment]
Offer still available.

**Comment by alexiordan** [comment]
I got the `System.Runtime.InteropServices.SEHException` again. This time only after about 45 minutes of server downtime.

I was just reconnecting (not re-build the HubConnection).

I also save a process dump or it. How can I send it?

Some info (there is some info after the below table):

  | Name | Value | Type
-- | -- | -- | --
◢ | $exception | {System.Runtime.InteropServices.SEHException (0x80004005): External component has thrown an exception.} | System.Runtime.InteropServices.SEHException
  | ▶ Data | {System.Collections.ListDictionaryInternal} | System.Collections.IDictionary {System.Collections.ListDictionaryInternal}
  | ErrorCode | 0x80004005 | int
  | HResult | 0x80004005 | int
  | HelpLink | null | string
  | ▶ IPForWatsonBuckets | {0} | System.UIntPtr
  | ▶ InnerException | null | System.Exception
  | IsTransient | false | bool
  | Message | "External component has thrown an exception." | string
  | RemoteStackTrace | null | string
  | Source | null | string
  | StackTrace | null | string
  | TargetSite | null | System.Reflection.MethodBase
  | WatsonBuckets | null | object
  | _HResult | 0x80004005 | int
  | _className | null | string
  | ▶ _data | {System.Collections.ListDictionaryInternal} | System.Collections.IDictionary {System.Collections.ListDictionaryInternal}
  | _dynamicMethods | null | object
  | _exceptionMethod | null | System.Reflection.MethodBase
  | _exceptionMethodString | null | string
  | _helpURL | null | string
  | ▶ _innerException | null | System.Exception
  | ▶ _ipForWatsonBuckets | {0} | System.UIntPtr
  | _message | "External component has thrown an exception." | string
  | _remoteStackIndex | 0x00000000 | int
  | _remoteStackTraceString | null | string
  | _source | null | string
  | _stackTrace | null | object
  | _stackTraceString | null | string
  | _watsonBuckets | null | object
  | _xcode | 0xe0434352 | int
  | ▶ _xptrs | {0} | System.IntPtr
  | ▶ Static members |   |  

Looking after that ErrorCode (0x80004005) on my saved error logs, I've seen it in the past in following shape:
```
Unable to connect to the remote server --> System.Net.WebSockets.WebSocketException (0x80004005): Unable to connect to the remote server ---> System.Exception: CannotConnect ---> System.Runtime.InteropServices.COMException: The text associated with this error code could not be found.

A connection with the server could not be established
```

I let processes running again so if you need any additional information, just ask.

**Comment by anurse** [comment]
The dump _may_ be useful but given the lack of a stack trace in your exception I'm not sure how useful it will be. Since I'd imagine it's fairly large, you would need to upload it to some file sharing service (OneDrive, DropBox, etc.) and send me a link (preferably over email, I see you've already reached out over email).

The lack of a stack trace will make this very difficult to diagnose though. Can you verify that you have `try...catch` blocks wrapping all your `StartAsync`, `InvokeAsync`, `SendAsync`, etc. calls on `HubConnection`? In theory, any of them could throw this error if the server goes down, and if it goes unhandled, it appears UWP doesn't preserve the stack trace properly.

**Comment by alexiordan** [comment]
OK.
I've sent a link to dump.
They are all caught.

But I will try again maybe next time will have a proper stack trace.

**Comment by anurse** [comment]
Can you clarify if you are seeing these exceptions crashing your app or if they are just appearing in the Uncaught Exception event? These exceptions would likely be thrown from one of our background threads, but they shouldn't be crashing your application. I just want to get an idea of what you are looking to change here. If the server is offline, obviously SignalR will have failures connecting to the server.

**Comment by alexiordan** [comment]
The final exception, SEHException it is crashing the app.

I am not sure I mentioned that I am running the app on a Dragonboard 410C, if it makes any difference.

**Comment by icistrate** [comment]
I am also having issues trying to reconnect. 

@anurse Simply calling hubConnection.StartAsync() does not work. I tried using a timer to handle reconnection, but the behavior is very strange: it looks like parallel ghost connections are awaken on StartAsync. So, as I see it, recreating the connection is probably a must. 

Would it be possible to make the HubConnection to cleanup its internal state during "StartAsync" call?

**Comment by anurse** [comment]
@icistrate can you file a new issue with a sample to reproduce the issue you're seeing? If you can show what you mean by ghost connections (memory traces, network statistics, etc.) that would help. We'd also need client logs as described by the [Diagnostics Guide](https://github.com/aspnet/SignalR/wiki/Diagnostics-Guide) to see if we can track this down.

> Would it be possible to make the HubConnection to cleanup its internal state during "StartAsync" call?

HubConnection cleans up it's internal state when stopped, either due to the connection being disconnected by the server, or when `StopAsync` is called. If you're seeing other behavior, we definitely want to know about that.

**Comment by radenkozec** [comment]
@alexiordan I manage to create working version. You can check out my blog post about it here:
https://www.radenkozec.com/net-core-signalr-automatic-reconnects/

**Comment by radenkozec** [comment]
@alexiordan Regarding Disposing I got tip from @anurse that I don't need to call dispose.
I test it only for few minutes and it works without a issue.
Did not done extensive testing.
I hope you and some other guys will also test it and leave me your feedback.

**Comment by alexiordan** [comment]
For minutes it works for me too :-)
I mainly do the same thing as your code does this is why I am curious.
But as soon as I will have any results I'll post them here.

**Comment by icistrate** [comment]
In my implementation (using a simple timer below) I also had to DisposeAsync the connection, else I would have new set of callbacks listening on the connection each time StartAsync is executed.

@radenkozec Thanks for the tips of using Polly.. hope I will find the time to improve my code one day.

**Comment by radenkozec** [comment]
@alexiordan Did not seen that I don't need to recreate HubConnection. Might work need to test that.
@icistrate Here I got that I don't need to call DisposeAsync https://github.com/aspnet/SignalR/issues/2845

**Comment by anurse** [comment]
> I remember trying to use StopAsync and still getting that strange behavior with the non-disposed increasing callbacks.

I assume you found that by using a memory profile? We don't see that behavior in our testing, can you provide a sample that reproduces the issue and post it in a new issue so we can investigate? Calling `StopAsync` should release all per-connection resources, so SignalR's memory consumption shouldn't be increasing with each restart.

**Comment by radenkozec** [comment]
@alexiordan @icistrate I updated my code to re-use HubConnection.

**Comment by simeyla** [comment]
Why can't we just get some reasonable / simple built-in functionality to do this?

      hubConnectionBuilder.withReconnectStrategy({ reconnect: true, maxReconnectTime: 20, hardFailureCallback: () => ... });

Surely an option to HubConnectionBuilder with some simple options wouldn't take much and would lead to much better out of the box performance. 

This is exactly the type of project where trusting people to do it right is going to backfire! Especially when using Azure SignalR I would have thought that some well considered heuristics would be essential to avoid a 'reconnect storm' or accidental infinite loops causing a 'reconnect tornado' from a single client.

**Comment by bradygaster** [comment]
@simeyla i really like that idea, frankly, of providing a strategy for reconnections. See issue #1057 as it is marked with a 3.0.0 milestone so we're considering it for our next iteration. I like your signature above, though, it is an interesting approach to solving the issue that's relatively consistent with the other methods that hang from hubConnectionBuilder.

**Comment by davidfowl** [comment]
Yes, we think it's important to implement something since this scenario is so common. We didn't want to rush the implementation in the first version since reconnect has been a large source of problems in the previous version of SignalR.
